Counts to radians

PAJohnson opened the pull request:
This changes the units of Controller from counts to radians wherever counts was used. As part of this change, pos_est_rad_, vel_est_rad_, and pos_cpr_rad_ are now available from Encoder. Gains have been updated where appropriate and HWIL tests pass on my test rig.

PAJohnson committed:
Added "shadow" variables for counts_to_rads. Controller::update() tested and working with radian inputs.

PAJohnson committed:
Changed from counts->radians internally. By default, input positions and velocities are also in radians.
Added multipliers for input pos/vel to allow users to define their own units for those inputs.

PAJohnson committed:
Modified tests for counts->rad conversion, all passing on my test rig (PJ)

PAJohnson committed:
Removed unit-conversion abiltiy for input_pos and input_vel
Updated documentation to reflect unit change.

PAJohnson committed:
Removed commented line in controller.hpp
Added missing ')' to closed_loop_test.py
Invalid enum definitions in uart_ascii_test.py fixed.

Wetmelon commented on Firmware/MotorControl/controller.hpp:
```diff
@@ -95,8 +95,6 @@ class Controller : public ODriveIntf::ControllerIntf {
     bool anticogging_valid_ = false;
 
-    // custom setters
-    void set_input_pos(float value) { input_pos_ = value; input_pos_updated(); }
```
Did this get moved somewhere?  The trap_traj planner needs input_pos_updated() called so that it can plan a new trajectory.

PAJohnson replied:
Good catch, I added it back.

Wetmelon commented on Firmware/MotorControl/controller.cpp:
```diff
@@ -128,19 +128,19 @@ bool Controller::update(float* torque_setpoint_output) {
     // TODO also enable circular deltas for 2nd order filter, etc.
     if (pos_wrap_src_) {
-        float cpr = *pos_wrap_src_;
+        float cpr = *pos_wrap_src_ * 2.0f * M_PI / ((float)axis_->encoder_.config_.cpr);
```
I'm not sure the float cast is needed here, but in general prefer static_cast<float>() to (float)()

PAJohnson replied:
I removed the C-style casts I added if it would have been implicitly done anyway.

madcowswe replied:
I disagree. I think the (float) are good for clarity that we are converting int to float here. But `static_cast` is too verbose and hard to read. All this code is on POD anyway, so there is no problem using (float).

madcowswe replied:
I put the `(float)'s` back.

PAJohnson committed:
Un-removed custom setter for input_pos_
removed explicit casts where variables would be promoted anyway

PAJohnson committed:
Merge branch 'devel' of https://github.com/madcowswe/ODrive into counts_to_rads

Merging recent devel changes into branch

madcowswe committed:
Revert "Un-removed custom setter for input_pos_"

This reverts commit 744bfeb110d87b5caa534ac921ef9cad2654e772.

madcowswe committed:
un-remove custom setter for input-pos

madcowswe commented on Firmware/MotorControl/controller.hpp:
```diff
@@ -95,8 +95,6 @@ class Controller : public ODriveIntf::ControllerIntf {
     bool anticogging_valid_ = false;
 
-    // custom setters
-    void set_input_pos(float value) { input_pos_ = value; input_pos_updated(); }
```
Looks to me like it wasn't put back yet (only restored in the yaml, this implementation is still missing)

PAJohnson replied:
Looks like I missed a commit here. Resolved.

madcowswe commented on Firmware/MotorControl/controller.hpp:
```diff
@@ -77,15 +77,15 @@ class Controller : public ODriveIntf::ControllerIntf {
-    float input_torque_ = 0.0f;
+    float input_pos_ = 0.0f;            // [radians]
+    float input_vel_ = 0.0f;            // [rad/s]
+    float input_torque_ = 0.0f;         // [Nm]
```
The units comments looks a bit weird flying out to the right so much...

PAJohnson replied:
Fixed in 7ce896c

madcowswe commented on Firmware/MotorControl/encoder.hpp:
```diff
@@ -84,6 +84,10 @@ class Encoder : public ODriveIntf::EncoderIntf {
 
+    float pos_est_rad_ = 0.0f;  // [rad]
+    float vel_est_rad_ = 0.0f;  // [rad]
+    float pos_cpr_rad_ = 0.0f;  // [rad]
```
let's rename xxx_estimate -> xxx_estimate_counts, and then rename xxx_est_rad -> xxx_estimate.

While the counts based one is the "core" one that pertains to the encoder, and from that perspective it makes sense that it's the base one; from the user's perspective it should be the radians one that is core and the counts one just for information.

PAJohnson replied:
Fixed in 7ce896c

madcowswe commented on Firmware/MotorControl/encoder.cpp:
```diff
@@ -522,6 +522,11 @@ bool Encoder::update() {
         snap_to_zero_vel = true;
     }
 
+    //expose pos/vel estimates in radians for Controller
```
IMO it shoudl be semantically the case that the "main output" of the Encoder is results in radians, and the count based variables are kind of an internal implementation detail (still ok to expose it of course).

PAJohnson replied:
Fixed in 7ce896c

PAJohnson committed:
Merge branch 'devel' of https://github.com/madcowswe/ODrive into counts_to_rads

Pull devel into counts_to_rads

PAJohnson committed:
In encoder.*pp, XXX_estimate -> XXX_estimate_counts and XXX_est_rad -> XXX_estimate.
Changed HWIL tests to reflect unit change
Corrected sources for position and velocity estimates elsewhere.

PAJohnson committed:
Merge branch 'counts_to_rads' of https://github.com/madcowswe/ODrive into counts_to_rads

PAJohnson committed:
Changed conditions and sources of wrapping for the circular setpoint feature as it related to this unit change. setpoints_in_cpr is still not exposed as an endpoint due to implementation changes when input_mode was introduced.

madcowswe commented on Firmware/MotorControl/controller.hpp:
```diff
@@ -21,15 +21,16 @@ class Controller : public ODriveIntf::ControllerIntf {
+        float vel_ramp_rate = 2.0f * M_PI;              // [(rad/s) / s]
         float torque_ramp_rate = 0.01f;                 // Nm / sec
         bool setpoints_in_cpr = false;
+        float circular_setpoint_range = 2.0f * M_PI;    //circular space if setpoints_in_cpr is used for controller
```
Our convention is to have a space between the `//` and the comment. Usually also start with capital letter.

PAJohnson replied:
Fixed in a262688

madcowswe commented on Firmware/MotorControl/controller.hpp:
```diff
@@ -75,17 +76,17 @@ class Controller : public ODriveIntf::ControllerIntf {
     float* vel_estimate_src_ = nullptr;
     bool* vel_estimate_valid_src_ = nullptr;
-    int32_t* pos_wrap_src_ = nullptr; // enables circular position setpoints if not null. The value pointed to is the maximum position value.
+    float* pos_wrap_src_ = nullptr; // enables circular position setpoints if not null. The value pointed to is the maximum position value.
```
"enables circular position setpoints if not null" This logic was changed, you can probably just delete that part.

PAJohnson replied:
Fixed in a262688

madcowswe commented on Firmware/MotorControl/controller.hpp:
```diff
@@ -96,7 +97,8 @@ class Controller : public ODriveIntf::ControllerIntf {
 
     // custom setters
-    void set_input_pos(float value) { input_pos_ = value; input_pos_updated(); }
+    void set_input_pos(float value) { input_pos_ = value; input_pos_updated();}
```
formatting: `input_pos_updated();}` -> `input_pos_updated(); }`

PAJohnson replied:
Fixed in a262688

madcowswe commented on Firmware/MotorControl/encoder.hpp:
```diff
@@ -74,16 +74,20 @@ class Encoder : public ODriveIntf::EncoderIntf {
 
+    float pos_estimate_ = 0.0f;  // [rad]
+    float vel_estimate_ = 0.0f;  // [rad]
+    float pos_cpr_ = 0.0f;  // [rad]
```
It used to be the case that circular mode was between 0 and `cpr`, because `cpr` was "one turn", i.e. circular mode is within one turn. But now "in one turn" no longer has anything to do with `cpr`.
I would suggest we rename this to `pos_circular_`. Similarly we should rename `setpoints_in_cpr` and other things like that (if any (I forget if there are more)).

PAJohnson replied:
Fixed in commit a262688

madcowswe commented on docs/control.md:
```diff
@@ -34,15 +34,15 @@ For more detail refer to [controller.cpp](https://github.com/madcowswe/ODrive/bl
 The ultimate output of the controller is the voltage applied to the gate of each FET to deliver current through each coil of the motor. The current through the motor linearly relates to the torque output of the motor. This means that the inputs to the cascaded controller are theoretically the position (angle), velocity (angle/time), and acceleration (angle/time/time) of the motor. Note that when  ...
 
-The output of each stage of the controller is clamped before being fed into the next stage. So after the `vel_cmd` is calculated from the position controller, the `vel_cmd` is clamped to the velocity limit. The `current_cmd` output of the velocity controller is then clamped and fed to the current controller. Oddly enough the controller class does not contain the current controller, but instead th ...
+The output of each stage of the controller is clamped before being fed into the next stage. So after the `vel_cmd` is calculated from the position controller, the `vel_cmd` is clamped to the velocity limit. The `torque_cmd` output of the velocity controller is then clamped and fed to the motor controller. Oddly enough the controller class does not contain the motor controller, but instead the mot ...
```
I don't think it makes sense to say `motor controller` instead of `current controller` here...

PAJohnson replied:
Fixed in commit a262688

madcowswe commented on docs/control.md:
```diff
@@ -34,15 +34,15 @@ For more detail refer to [controller.cpp](https://github.com/madcowswe/ODrive/bl
+The output of each stage of the controller is clamped before being fed into the next stage. So after the `vel_cmd` is calculated from the position controller, the `vel_cmd` is clamped to the velocity limit. The `torque_cmd` output of the velocity controller is then clamped and fed to the motor controller. Oddly enough the controller class does not contain the motor controller, but instead the mot ...
 
-The feedforward terms available when using the position or velocity control mode are meant to enable better performance when the dynamics of a system are known and the host controller can predict the motion based on the load. A perfect example of this is the use of the trajectory controller that sets the position, velocity, and current based on the desired position, velocity, and acceleration. If ...
+The feedforward terms available when using the position or velocity control mode are meant to enable better performance when the dynamics of a system are known and the host controller can predict the motion based on the load. A perfect example of this is the use of the trajectory controller that sets the position, velocity, and current based on the desired position, velocity, and acceleration. If ...
```
"position, velocity, and current" -> "position, velocity, and torque"

PAJohnson replied:
Fixed in commit a262688

PAJohnson committed:
Merge branch 'devel' of https://github.com/madcowswe/ODrive into counts_to_rads

PAJohnson committed:
Added pos_circular_ to Encoder for tracking position in a circular space
Corrected docs as they relate to unit changes
Added pos_estimate_linear and pos_estimate_circular to Controller and eliminated pointer switching logic in Controller::select_encoder()

PAJohnson committed:
Changed old-school variable reuse to modern style declarations

PAJohnson committed:
Changed from radians internally to units of "turns" where a full rotation is 1.0 turn
Changed documentation to reflect the unit change

madcowswe committed:
Fix phase_vel units

madcowswe committed:
Add _src to estimate pointers

madcowswe committed:
Merge branch 'devel' into counts_to_rads

madcowswe committed:
Also add _src in header

madcowswe commented on Firmware/MotorControl/axis.cpp:
```diff
@@ -319,9 +320,9 @@ bool Axis::run_closed_loop_control_loop() {
-    // TODO: Also do this for circular position mode
-    controller_.pos_setpoint_ = *controller_.pos_estimate_src_;
-    controller_.input_pos_ = *controller_.pos_estimate_src_;
+    // TODO: use circular src if in circular mode.
```
Can you fix this TODO? I think there is a similar situation in `run_homing` also.

madcowswe commented on Firmware/MotorControl/controller.cpp:
```diff
@@ -139,10 +137,9 @@ bool Controller::update(float* torque_setpoint_output) {
     // TODO also enable circular deltas for 2nd order filter, etc.
-    if (pos_wrap_src_) {
-        float cpr = *pos_wrap_src_;
+    if (config_.circular_setpoints && pos_estimate_circular) {
```
is `pos_estimate_circular` needed here? it's not used in the body.

madcowswe commented on Firmware/MotorControl/controller.cpp:
```diff
@@ -229,20 +226,19 @@ bool Controller::update(float* torque_setpoint_output) {
     if (config_.control_mode >= CONTROL_MODE_POSITION_CONTROL) {
         float pos_err;
-        if (!pos_estimate_src) {
+        if (!pos_estimate_linear || !pos_estimate_circular) {
```
Probably put two separate error checks in each branch of the below `if (config_.circular_setpoints ...`

madcowswe commented on Firmware/MotorControl/controller.cpp:
```diff
@@ -229,20 +226,19 @@ bool Controller::update(float* torque_setpoint_output) {
 
-        if (pos_wrap_src_) {
-            float cpr = *pos_wrap_src_;
+        if (config_.circular_setpoints && pos_estimate_circular) {
```
Check for `pos_estimate_circular` redundant (due to error check).

madcowswe committed:
Update controller.hpp

madcowswe commented on Firmware/MotorControl/trapTraj.hpp:
```diff
@@ -4,9 +4,9 @@
-        float decel_limit = 5000.0f; // [count/s^2]
+        float vel_limit = 2.0f;  // [turn/s]
+        float accel_limit = 0.5f;       // [turn/s^2]
+        float decel_limit = 0.5f;       // [turn/s^2]
```
comments are flying out to the right.

madcowswe commented on Firmware/MotorControl/encoder.cpp:
```diff
@@ -132,7 +132,7 @@ void Encoder::set_circular_count(int32_t count, bool update_offset) {
 
     // Update states
     count_in_cpr_ = mod(count, config_.cpr);
-    pos_cpr_ = (float)count_in_cpr_;
+    pos_cpr_counts_ = (float)count_in_cpr_;
```
`Encoder::set_circular_count` is updating `pos_cpr`, but the expectation is probably that it should update `pos_circular`. Perhaps we should drop `pos_cpr` and make only `pos_circular` exist?

madcowswe replied:
We can drop `pos_cpr` after this PR.

PAJohnson replied:
I agree - I will leave `set_circular_count()` how it is for now and I can fix it after this PR.

madcowswe commented on docs/commands.md:
```diff
@@ -44,8 +44,8 @@ Possible values are listed [here](api/odrive.axis.controller.controlmode).
-* `<axis>.controller.input_pos = <encoder_counts>`
-* `<axis>.controller.input_vel = <encoder_counts/s>`
+* `<axis>.controller.input_pos = <radians>`
+* `<axis>.controller.input_vel = <radians/s>`
```
turns, turns/s

madcowswe commented on docs/getting-started.md:
```diff
@@ -250,7 +250,7 @@ Let's get motor 0 up and running. The procedure for motor 1 is exactly the same,
 
 2. Type `odrv0.axis0.requested_state = AXIS_STATE_CLOSED_LOOP_CONTROL` <kbd>Enter</kbd>. From now on the ODrive will try to hold the motor's position. If you try to turn it by hand, it will fight you gently. That is unless you bump up `odrv0.axis0.motor.config.current_lim`, in which case it will fight you more fiercely. If the motor begins to vibrate either immediately or after being disturbed yo ...
-3. Send the motor a new position setpoint. `odrv0.axis0.controller.input_pos = 10000` <kbd>Enter</kbd>. The units are in encoder counts.
+3. Send the motor a new position setpoint. `odrv0.axis0.controller.input_pos = 10` <kbd>Enter</kbd>. The units are in radians.
```
turns.

madcowswe commented on docs/getting-started.md:
```diff
@@ -273,7 +273,7 @@ Asking the ODrive controller to go as hard as it can to raw setpoints may result
 Set the filter bandwidth: `axis.controller.config.input_filter_bandwidth = 2.0` [1/s]<br>
 Activate the setpoint filter: `axis.controller.config.input_mode = INPUT_MODE_POS_FILTER`.<br>
-You can now control the velocity with `axis.controller.input_pos = 1000` [counts].
+You can now control the velocity with `axis.controller.input_pos = 10` [radians].
```
turns

madcowswe commented on docs/getting-started.md:
```diff
@@ -294,8 +294,8 @@ In the above image blue is position and orange is velocity.
-`accel_limit` is the maximum acceleration in counts / sec^2<br>
-`decel_limit` is the maximum deceleration in counts / sec^2<br>
+`accel_limit` is the maximum acceleration in radians / sec^2<br>
+`decel_limit` is the maximum deceleration in radians / sec^2<br>
```
turns

madcowswe commented on docs/interfaces.md:
```diff
@@ -113,14 +113,14 @@ You can control the ODrive directly from an hobby RC receiver.
 
 Any of the numerical parameters that are writable from the ODrive Tool can be hooked up to a PWM input.
-As an example, we'll configure GPIO4 to control the angle of axis 0. We want the axis to move within a range of -1500 to 1500 encoder counts.
+As an example, we'll configure GPIO4 to control the angle of axis 0. We want the axis to move within a range of -2 to 2 radians.
```
turns

madcowswe commented on docs/odrivetool.md:
```diff
@@ -216,8 +216,8 @@ For example, to plot the approximate motor torque [N.cm] and the velocity [RPM]
         cancellation_token = start_liveplotter(lambda: [
-            (((my_odrive.axis0.encoder.pll_vel)/8192)*60), # 8192 CPR encoder
-            ((8.27*my_odrive.axis0.motor.current_control.Iq_setpoint/150) * 100), # Torque [N.cm] = (8.27 * Current [A] / KV) * 100
+            ((my_odrive.axis0.encoder.vel_estimate*60/(6.2832)), # radians to rpm
```
Update calculation as we now use turns instead of rad.

madcowswe commented on tools/odrive/shell.py:
```diff
@@ -32,7 +32,7 @@ def print_help(args, have_devices):
     print('For example: "odrv0.motor0.encoder.pos_estimate"')
     print('will print the current encoder position on motor 0')
-    print('and "odrv0.motor0.pos_setpoint = 10000"')
+    print('and "odrv0.motor0.input_pos = 10000"')
```
10k is too large now.

PAJohnson committed:
Fixed null pointer checks in controller.cpp
Fixed formatting in trapTraj.hpp
Fixed how pos_setpoint is set in axis.cpp to avoid transients on startup so that it will handle circular position setpoints
More unit corrections in the docs: radians -> turns

PAJohnson committed:
Merge branch 'devel' of https://github.com/madcowswe/ODrive into counts_to_rads

Merge devel into counts_to_rads (turns) - conflict for the step input callback

PAJohnson committed:
In axis: counts_per_step to turns_per_step (broken during previous merge)

PAJohnson committed:
Documentation fixes for unit change
Fixed naming, docs, and test for CAN protocol - set trap traj a per css is now set trap traj inertia

PAJohnson committed:
Fixed units for inertia

PAJohnson committed:
Fixed units for can_simple protocol and updated can HWIL test

PAJohnson committed:
Added closed loop test to CAN HWIL test
Fixed ascii protocol docs: current -> torque

PAJohnson committed:
Split closed loop test from can_test.py into integration_test.py

madcowswe commented on Firmware/MotorControl/axis.cpp:
```diff
@@ -320,9 +318,14 @@ bool Axis::run_closed_loop_control_loop() {
-    // TODO: use circular src if in circular mode.
-    controller_.pos_setpoint_ = *controller_.pos_estimate_linear_src_;
-    controller_.input_pos_ = *controller_.pos_estimate_linear_src_;
+    if(controller_.config_.circular_setpoints == true) {
```
style: add space `if (stuff`

madcowswe replied:
style: please use `if(bool)` not `if(bool = true)`.

madcowswe commented on Firmware/MotorControl/axis.cpp:
```diff
@@ -320,9 +318,14 @@ bool Axis::run_closed_loop_control_loop() {
+    }
+    else {
+        controller_.pos_setpoint_ = *controller_.pos_estimate_linear_src_;
+        controller_.input_pos_ = *controller_.pos_estimate_linear_src_;
```
No check for Null ptrs.

madcowswe commented on Firmware/MotorControl/axis.cpp:
```diff
@@ -372,8 +375,13 @@ bool Axis::run_homing() {
+    }
+    else {
+        controller_.pos_setpoint_ = *controller_.pos_estimate_linear_src_;
+    }
```
(same as above): space, compare to true on `if`; check null ptr

PAJohnson committed:
Added null pointer checks to axis.cpp for pos_estimate_X
Fixed formatting
Updated CHANGELOG.md

PAJohnson committed:
Better style for null pointer checks
